README.md
# interval-conversions

Parses user strings into ms intervals, and makes configurable user-friendly strings out of ms intervals.

It is faithful with regards to the variable durations of years and months. For example, starting from 2000-01-02 (February of a leap year), 1 month means 29 days, and -1 month means -31 days.

This is an ESM package. If your project is ESM as well, there will be no issues. But it cannot be used with CommonJS's `require()`. [This write-up](https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c) goes into some ways to deal with this, and some problems you might run into. Note that option 3 does not apply, since this package has always been ESM.

Since this package is so simple, I am open to regressing to CommonJS for compatibility, but I don't know how to do this while ensuring I don't break any existing usage.

## Parse interval
```ts
parseInterval(text: string, startDate?: Date): number | undefined
const milliseconds = parseInterval("5d"); // 432000000
```
Parses user input like "5y6mo" and "3.5 minutes" into ms intervals. It is fairly flexible, but still expects the user to follow its formatting rules. It makes no attempt at interpreting malformed strings. It will return undefined if it could not parse the text. 

The actual format is as follows:

`<number> years <number> months <number> weeks <number> days <number> hours <number> minutes <number> seconds`

Each unit is optional, but all present units need to be in order. All units are case insensitive. All spaces are optional. Numbers for years and months can't have decimals, but those for the other units can. The combination of units does not need to make sense, e.g. "1 week 20 days" will simply be 27 days, and ".5d12h" will simply be 1 day.

* `years` can also be written as `year` or `y`
* `months` can also be written as `month` or `mo`
* `weeks` can also be written as `week` or `w`
* `days` can also be written as `day` or `d`
* `hours` can also be written as `hour`, `hrs`, `hr` or `h`
* `minutes` can also be written as `minute`, `mins`, `min` or `m`
* `seconds` can also be written as `second`, `secs`, `sec` or `s`

A `-` can be inserted before any number to subtract all the units that follow it. Another `-` will make the following units additive again (as if subtracting from the previous subtraction). For example, "1d - 10m 30s" describes an interval 10.5 minutes short of a day. "1d - 10m -30s" describes an interval 9.5 minutes short of a day. Intervals can be negative as a whole, resulting in a negative number output.

Because years and months vary in their exact ms duration, a Date object can optionally be passed to be used as a starting point. If no Date object is provided, it will make one representing the now.

## Stringify interval
```ts
stringifyInterval(interval: number, options?: Date | StringifyOptions): string
const text = stringifyInterval(50000000000); // "578 days, 16 hours and 53 minutes"
const text2 = stringifyInterval(50000000000, new Date("1950")); // "1 year, 7 months, 1 day, 16 hours and 53 minutes"
```
Generates a user-friendly interval description from a ms interval. It is formatted like "1 day, 5 hours and 20 minutes". By default, if the total duration is under 10 minutes, it will say seconds too. The second argument can be either a `Date`, or a `StringifyOptions` object. The `StringifyOptions` object has an optional `startDate` property that is a `Date`. If a date is supplied by either means, it will by default say years and months, if appropriate, using the date as a starting point. Negative intervals will go backwards from the starting point, positive ones forwards. Without the date, it will not say months or years, no matter the interval or the options. A `NaN` will result in an empty string.

The full `StringifyOptions` with all values explicitly set to their defaults is as follows:
```ts
{
	startDate: undefined,
	thresholds: {
		years: [0, Infinity], // On
		months: [0, Infinity], // On
		weeks: [Infinity, 0], // Off
		days: [0, Infinity], // On
		hours: [0, Infinity], // On
		minutes: [0, Infinity], // On
		seconds: [0, 600] // On until number exceeds 600 seconds (10 minutes)
	},
	pad: {
		years: false,
		months: false,
		weeks: false,
		days: false,
		hours: false,
		minutes: false,
		seconds: false,
	},
	displayZero: {
		years: false,
		months: false,
		weeks: false,
		days: false,
		hours: false,
		minutes: false,
		seconds: false,
	},
	strings: {
		years: ["year", "years"],
		months: ["month", "months"],
		weeks: ["week", "weeks"],
		days: ["day", "days"],
		hours: ["hour", "hours"],
		minutes: ["minute", "minutes"],
		seconds: ["second", "seconds"],
		spacer: " ",
		joiner: ", ",
		finalJoiner: " and "
	},
}
```
Each property is optional, and so is each property of those properties. Below each property is explained.

### Avoiding re-processing options

Processing provided `StringifyOptions` is a relatively expensive operation. In general, the more types of options set, the longer it takes to process. Under some circumstances it can be more than half of the total time stringifying the interval. That is why there is a class `Stringifier`. To avoid unnecessary re-processing, `Stringifier` can be instantiated with a `StringifierOptions` object. `StringifierOptions` is just `StringifyOptions` minus `startDate`. The resulting object has a `stringify` method that optionally takes a `Date` as starting date.

When sticking to the defaults, using `Stringifier` should be unnecessary.

Note that here, a `NaN` interval will throw an error instead of returning an empty string.

```ts
for (let i = 0; i < 10; i++) { // Wasteful, processing options 10 times in a row
	const text = stringifyInterval(500000, {
		startDate: new Date(),
		pad: true,
		strings: { minutes: "mins", seconds: "secs" },
	}); // 08 mins and 20 secs
}
const stringifier = new IntervalStringifier({
	pad: true,
	strings: { minutes: "mins", seconds: "secs" },
});
for (let i = 0; i < 10; i++) { // Efficient, processing options only once, then using it 10 times
	const text = stringifier.stringify(500000, new Date()); // 08 mins and 20 secs
}
```

### Custom thresholds

The `thresholds` property is an object with optional time unit properties. Each property needs to be `[number, number]`, a number or a boolean, or be left undefined. If it is `[number, number]`, the first number is the lower threshold (expressed in that unit) for the unit to appear, and the second number is the upper threshold beyond which it no longer appears. If it is a number, it will be treated as the upper threshold, with the lower threshold set to 0. If it is `true`, it will be treated like `[0, Infinity]`, and if it is `false`, it will be treated like `[Infinity, 0]` (technically there are countless ways to express an unreachable threshold, I picked this one). `Infinity` is a valid value for a threshold.

Fractional parts of thresholds for years and months do nothing.

Even at a lower threshold of 0, a unit will by default not appear if there are zero of that unit. But if smaller units are disabled, the interval may be rounded up to have 1 of that unit.

Note that it is possible and easy to provide settings that don't produce a sensible result. For example, some ranges of inputs could end up without units at all (resulting in an empty string), or years could disappear and be converted to months when there are too many. It is up to whoever configures the thresholds to make sure they make sense.

#### Example: single unit

The following configuration will make it only ever output one unit. Weeks could also be enabled with `[0, 4]` with days set to `[0, 7]`. Note that due to the variability of months, it is not currently possible to make days go to 29 and 30, without also making it sometimes output something like "1 month and 1 day". Also note that the upper limit of months has to be 11, rather than 12, due to the different way rounding works for variable duration units (years and months).
```ts
const oneUnit: StringifyThresholds = {
	years: [0, Infinity],
	months: [0, 11],
	weeks: false,
	days: [0, 28],
	hours: [0, 24],
	minutes: [0, 60],
	seconds: [0, 60]
};
const text = stringifyInterval(interval, { thresholds: oneUnit });
```

### Padding with 0s and displaying even when 0

The `pad` property controls whether to pad unit values. Unit values will be padded with 0s at the start to ensure they are at least 2 characters, except for year values, where it will pad to at least 4 characters.

The `displayZero` property controls whether a unit shows up when it falls within the threshold, even when the value is 0.

Both of these can set to `true`, meaning all enabled, or `false` meaning all disabled (which is also the default). But they can also be passed an object with individual time units (like in `thresholds`) set to `true` or `false`.

#### Example: displaying 0 values
```ts
const text1 = stringifyInterval(parseInterval("10m"), { displayZero: true }); // "0 years, 0 months, 0 days, 0 hours, 10 minutes and 0 seconds"
const text2 = stringifyInterval(parseInterval("3d2h1s"), { displayZero: true }); // "0 years, 0 months, 3 days, 2 hours, 0 minutes" (seconds were still omitted for exceeding the default upper threshold)
```

#### Example: a digital clock-like display

The following will implement an hh:mm:ss display. The first one has hours go potentially infinitely high, the second one displays days separately. The latter gets a little hacky with the strings, but it works.
```ts
const stringifierCompact = new IntervalStringifier({
	thresholds: { years: false, months: false, weeks: false, days: false, seconds: true },
	strings: { hours: "", minutes: "", seconds: "", joiner: ":", finalJoiner: ":", spacer: "" },
	pad: true,
	displayZero: true,
});
const text1 = stringifierCompact.stringify(parseInterval("3d2h1s")); // "74:00:01"
const text2 = stringifierCompact.stringify(parseInterval("1m")); // "00:01:00"
const stringifierCompact2 = new IntervalStringifier({
	thresholds: { years: false, months: false, weeks: false, days: true, seconds: true },
	strings: { days: [" day ", " days "], hours: ":", minutes: ":", seconds: "", joiner: "", finalJoiner: "", spacer: "" },
	pad: { hours: true, minutes: true, seconds: true },
	displayZero: { hours: true, minutes: true, seconds: true },
});
const text3 = stringifierCompact2.stringify(parseInterval("3d2h1s")); // "3 days 02:00:01"
const text4 = stringifierCompact2.stringify(parseInterval("1d12h34m56s")); // "1 day 12:34:56"
const text5 = stringifierCompact2.stringify(parseInterval("1y", new Date("2020")), new Date("2020")); // "366 days 00:00:00"
const text6 = stringifierCompact2.stringify(parseInterval("1m")); // "00:01:00"
```

### Custom strings

The `strings` property allows overriding the strings used to construct the output, for localization or whatever other reason. The time unit properties can be passed strings instead of arrays, in which case it will use the same string for the singular and the plural.

#### Example: condensed output
```ts
const shortStrings = {
	years: "y", months: "mo", weeks: "w", days: "d", hours: "h", minutes: "m", seconds: "s",
	spacer: "", joiner: " ", finalJoiner: " "
};
const text = stringifyInterval(-500000000, { strings: shortStrings }); // "5d 18h 53m"
```

### Typescript types

`StringifyOptions`, `StringifierOptions`, `StringifyThresholds`, `TimeUnitBooleans` and `StringSettings` are types exported to help construct those options objects in Typescript in a type safe way. They are relevant when wanting to construct the objects before passing them to the function or constructor.

## Stringify interval with single unit output

```ts
stringifyIntervalShort(interval: number, atLeast?: boolean): string
const text = stringifyIntervalShort(50000000000); // "2 years"
const text2 = stringifyIntervalShort(50000000000, true); // "1 year"
console.log(`The interval is somewhere between ${text2} and ${text}.`); // "The interval is somewhere between 1 year and 2 years."
```
Generates a user-friendly rough description of a ms interval, with only a single unit mentioned. It will pick the largest unit that contains the interval at least once, and then the smallest number that contains the interval. It is meant to be used in descriptions like "That event is within 2 years" or "The ban will expire in under 20 hours". It will use minutes, hours, days, months and years, but months and years are estimates based on average duration. It will prefix the number with `~` when it's not sure.

With the second argument as true, it will instead pick the largest number contained by the interval, to be used like "That event is at least 1 year away" or "The ban will last for at least 19 hours more".

CHANGELOG.md
## 1.4.0

- Allow padding time elements with zeroes.
- Allow displaying elements with a value of zero.
- Combined, those two things will allow output like 12:00:00.
- Added `IntervalStringifier`, a class that can be instantiated with a `StringifierOptions` object (which is just `StringifyOptions` minus `startDate`) to avoid repeated processing of options.

## 1.3.3

- Added "hrs" and "hr" as ways to specify the hours.

## 1.3.2

- Fixed an error in types that was causing some string options to incorrectly not be optional.
- Optimized a bit.

## 1.3.1

- Now exporting types `StringifyOptions`, `StringifyThresholds` and `StringSettings`, for help constructing named options objects in Typescript in a type safe way.

## 1.3.0

- `stringifyInterval`'s `stringifyOptions` can now have a `strings` property, allowing string overrides.

## 1.2.0

- `stringifyInterval` can now be passed a `stringifyOptions` object as the second argument. This can have a `startDate` property, which behaves the same as passing a `Date` object, and a `thresholds` property, which is an object with properties, all optional, to set the lower and upper thresholds between which each unit should be used.
- Fixed some rounding logic in `stringifyInterval` that could lead to things like `1 month and 31 days`.

## 1.1.0

- `stringifyInterval` can now be passed a `Date` object as the second argument, to get months and years in the output. The date will be used as a starting point, and negative or positive input matters.
- `stringifyInterval` no longer allows fractional digits in years or months. These already weren't working, but it quietly truncated them.
- `NaN` passed to `stringifyInterval` or `stringifyIntervalShort` will return an empty string. In a future version this will probably throw an error instead.


## 1.0.2

- A space between `-` and the following number is now allowed.

src/test/intense-tests.ts
import { stringifyInterval, StringifyOptions, StringifyThresholds, StringSettings } from "../stringifyInterval.js";

/*const oneUnit: StringifyThresholds = {
	years: [0, Infinity],
	months: [0, 11],
	weeks: false,
	days: [0, 28],
	hours: [0, 24],
	minutes: [0, 60],
	seconds: [0, 60]
};
const stepSize = -1000;


let lastOutcome = "";
let i = 0;
while (true) {
	const outcome = stringifyInterval(i, { startDate: new Date("2000"), thresholds: oneUnit });
	if (outcome !== lastOutcome) {
		console.log(`${outcome} from ${i}ms`);
		if (outcome.includes("2 years")) break;
		lastOutcome = outcome;
	}
	i += stepSize;
}*/

const possiblyExpensiveSettings: StringifyOptions = {
	startDate: new Date("2000"),
	thresholds: {
		years: false,
		months: false,
		weeks: false,
		days: true,
		seconds: true,
	},
	strings: {
		days: [" day ", " days "],
		hours: ":",
		minutes: ":",
		seconds: "",
		joiner: "",
		finalJoiner: "",
		spacer: "",
	},
	pad: {
		hours: true,
		minutes: true,
		seconds: true,
	},
	displayZero: {
		hours: true,
		minutes: true,
		seconds: true,
	},
};
const stepSize = -1000;

let lastOutcome = "";
let i = 0;
while (true) {
	const outcome = stringifyInterval(i, possiblyExpensiveSettings);
	if (outcome.slice(0, -8) !== lastOutcome.slice(0, -8)) {
		console.log(`${outcome} from ${i}ms`);
		if (outcome.startsWith("365 days")) break;
		lastOutcome = outcome;
	}
	i += stepSize;
}

src/parseInterval.ts
const regexInterval = /^(?:(?:(-) ?)?(\d+) ?y(?:ears?)?\s?)?(?:(?:(-) ?)?(\d+) ?mo(?:nths?)?\s?)?(?:(?:(-) ?)?(\d+(?:\.\d+)?|\.\d+) ?w(?:eeks?)?\s?)?(?:(?:(-) ?)?(\d+(?:\.\d+)?|\.\d+) ?d(?:ays?)?\s?)?(?:(?:(-) ?)?(\d+(?:\.\d+)?|\.\d+) ?h(?:(?:ou)?rs?)?\s?)?(?:(?:(-) ?)?(\d+(?:\.\d+)?|\.\d+) ?m(?:in(?:ute)?s?)?\s?)?(?:(?:(-) ?)?(\d+(?:\.\d+)?|\.\d+) ?s(?:ec(?:ond)?s?)?\s?)?$/i;

const SECOND = 1000;
const MINUTE = 60 * SECOND;
const HOUR = 60 * MINUTE;
const DAY = 24 * HOUR;
const WEEK = 7 * DAY;

/** Parses a user string like "1y 2mo 3w 4d 5h 6m 7s" and returns the duration in ms. If there is a component of variable duration (year or month), it will use the supplied date or now as a starting point. Returns undefined if parsing failed. */
export const parseInterval = (text: string, startDate?: Date): number | undefined => {
	text = text.trim();
	if (!text)
		return; // Empty input
	const result = regexInterval.exec(text);
	if (!result)
		return; // Didn't look like a duration
	let sign = 1;
	const parseNumber = (minus: string, number: string): number => {
		sign *= minus ? -1 : 1;
		return number ? sign * Number(number) : 0;
	};
	const timeElements = [];
	for (let i = 1; i < 15; i += 2) {
		timeElements.push(parseNumber(result[i], result[i + 1]));
	}
	const [y, mo, w, d, h, m, s] = timeElements;
	const invariable = w * WEEK + d * DAY + h * HOUR + m * MINUTE + s * SECOND;
	if (y === 0 && mo === 0) { // No variable elements
		return invariable;
	}
	const date = startDate ? new Date(startDate) : new Date();
	const startingPoint = date.getTime();
	date.setUTCFullYear(
		date.getUTCFullYear() + y,
		date.getUTCMonth() + mo,
	);
	return date.getTime() + invariable - startingPoint;
};

src/stringifyInterval.ts
const MINUTE_S = 60;
const HOUR_S = MINUTE_S * 60;
const DAY_S = HOUR_S * 24;
const WEEK_S = DAY_S * 7;

/** Processes the options (without startDate) once, which is a relatively expensive operation, then stores the processed configuration for future stringifier calls. */
export class IntervalStringifier {
	settings: StringifySettings;
	constructor(options: StringifierOptions = {}) {
		const { settings } = processStringifyOptions(options);
		this.settings = settings;
	}
	/**
	 * Stringifies a ms interval like "1 day, 5 hours and 20 minutes". If a date is supplied, it is capable of outputting years and months.
	 * 
	 * Behaves the same as stringifyInterval, except that all options (except startDate) were already processed by the constructor.
	 * 
	 * This one throws an error if the number is NaN, because that seems like a more proper outcome, and I got the opportunity to do that without changing existing API.
	 */
	stringify(interval: number, startDate?: Date): string {
		if (isNaN(interval))
			throw new RangeError("Cannot stringify NaN interval");
	
		return stringifyIntervalInternal(interval, this.settings, startDate);
	}
}

/** Stringifies a ms interval like "1 day, 5 hours and 20 minutes". If a starting date is supplied, it is capable of outputting years and months. */
export const stringifyInterval = (interval: number, options: Date | StringifyOptions = {}): string => {
	if (isNaN(interval))
		return "";

	const { settings, startDate } = processStringifyOptions(options);
	return stringifyIntervalInternal(interval, settings, startDate);
};

const stringifyIntervalInternal = (interval: number, settings: StringifySettings, startDate: Date | undefined): string => {
	interval = Math.round(interval / 1000);

	let inPast = false;
	if (interval < 0) {
		interval = -1 * interval;
		inPast = true;
	}

	const { thresholds, pad, displayZero, strings } = settings;

	const enabled: Record<TimeUnit, boolean> = {
		years: false,
		months: false,
		weeks: interval >= thresholds.weeks[0] * WEEK_S && interval <= thresholds.weeks[1] * WEEK_S,
		days: interval >= thresholds.days[0] * DAY_S && interval <= thresholds.days[1] * DAY_S,
		hours: interval >= thresholds.hours[0] * HOUR_S && interval <= thresholds.hours[1] * HOUR_S,
		minutes: interval >= thresholds.minutes[0] * HOUR_S && interval <= thresholds.minutes[1] * MINUTE_S,
		seconds: interval >= thresholds.seconds[0] && interval <= thresholds.seconds[1]
	};

	let shouldRoundToMonthsOrYears = false;
	[interval, shouldRoundToMonthsOrYears = false] = roundToSmallest(interval, enabled);

	let years = 0;
	let months = 0;
	let weeks = 0;
	let days = 0;
	let hours = 0;
	let minutes = 0;
	let seconds = 0;

	if (startDate && (thresholds.years[0] < Infinity || thresholds.months[0] < Infinity)) {
		[years, months, interval] = getYearsMonthsRemainder(startDate, interval, inPast, shouldRoundToMonthsOrYears, thresholds);
		enabled.years = years >= thresholds.years[0] && years <= thresholds.years[1];
		enabled.months = months + years * 12 >= thresholds.months[0] && months + years * 12 <= thresholds.months[1];
	}

	if (enabled.weeks) {
		weeks = Math.floor(interval / WEEK_S);
		interval -= weeks * WEEK_S;
	}
	if (enabled.days) {
		days = Math.floor(interval / DAY_S);
		interval -= days * DAY_S;
	}
	if (enabled.hours) {
		hours = Math.floor(interval / HOUR_S);
		interval -= hours * HOUR_S;
	}
	if (enabled.minutes) {
		minutes = Math.floor(interval / MINUTE_S);
		interval -= minutes * MINUTE_S;
	}
	if (enabled.seconds) {
		seconds = interval;
	} else {
		if (interval > 0 && (years > 0 || months > 0 || enabled.weeks || enabled.days || enabled.hours || enabled.minutes))
			console.warn("Some problem with the rounding in stringifyInterval");
	}


	let outputElements = 0;
	/** Inserts a comma or "and" (or nothing) as appropriate */
	const joiner = () => {
		outputElements--;
		if (outputElements > 1)
			return strings.joiner;
		if (outputElements === 1)
			return strings.finalJoiner;
		return "";
	};

	const timeElements: [TimeUnit, number][] = [
		["years", years],
		["months", months],
		["weeks", weeks],
		["days", days],
		["hours", hours],
		["minutes", minutes],
		["seconds", seconds]
	];
	// Calculate number of text elements to be joined with comma/and
	for (const [element, value] of timeElements) {
		if (enabled[element] && (displayZero[element] || value > 0)) outputElements++;
	}

	if (outputElements === 0) { // No elements, output 0 of smallest enabled
		for (const [element] of timeElements.reverse()) {
			if (enabled[element]) {
				return `0${strings.spacer}${strings[element][1]}`;
			}
		}
		//console.log(`stringifyInterval ended up with no enabled elements`);
		return ""; // Nothing enabled
	}

	let text = "";
	for (const [element, value] of timeElements) {
		if (enabled[element] && (displayZero[element] || value > 0)) {
			let numberPortion = value.toString();
			if (pad[element]) {
				const padCount = element === "years" ? 4 : 2;
				numberPortion = numberPortion.padStart(padCount, "0");
			}
			const elementString = value === 1 ? strings[element][0] : strings[element][1];
			text += `${numberPortion}${strings.spacer}${elementString}${joiner()}`;
		}
	}

	return text;
};

/**
 * For a given starting date and interval (and direction the interval should be interpreted as), returns the number of years and months, and the interval remaining after removing these years and months.
 * 
 * With shouldRound true, it also rounds to the smallest, enabled unit.
 * 
 * thresholds is used to determine which elements are enabled between the two.
 */
const getYearsMonthsRemainder = (startDate: Date, interval: number, inPast: boolean, shouldRound: boolean, thresholds: {
	years: [number, number];
	months: [number, number];
}): [number, number, number] => {
	const changingDate = new Date(startDate);
	const targetDate = new Date(changingDate.getTime() + interval * 1000 * (inPast ? -1 : 1)); // Date interval in the past or future

	const largerDate = inPast ? changingDate : targetDate;
	const smallerDate = inPast ? targetDate : changingDate;

	let years = largerDate.getUTCFullYear() - smallerDate.getUTCFullYear(); // Year difference
	let months = largerDate.getUTCMonth() - smallerDate.getUTCMonth(); // Month difference

	changingDate.setUTCFullYear(
		changingDate.getUTCFullYear() + (inPast ? -years : years),
		changingDate.getUTCMonth() + (inPast ? -months : months)
	); // Remove year and month difference
	if (largerDate.getTime() < smallerDate.getTime()) { // Went too far
		changingDate.setUTCMonth(changingDate.getUTCMonth() + (inPast ? 1 : -1));
		months--;
	}
	let remainingInterval = (largerDate.getTime() - smallerDate.getTime()) / 1000;
	if (months < 0) {
		months += 12;
		years--;
	} else if (months >= 12) {
		months -= 12;
		years++;
	}

	const enableYears = years >= thresholds.years[0] && years <= thresholds.years[1];
	const enableMonths = months + years * 12 >= thresholds.months[0] && months + years * 12 <= thresholds.months[1];

	if (!enableYears && !enableMonths) // Years and months ended up outside the threshold so it's simple
		return [0, 0, interval];

	if (!shouldRound) { // No rounding required, so still quite simple
		if (!enableYears) { // No years but there must still be months, so just push them into those
			months += years * 12;
			years = 0;
		} else if (!enableMonths) { // No months but there must still be years, so put the months back into interval
			const firstTime = changingDate.getTime();
			changingDate.setUTCMonth(changingDate.getUTCMonth() + (inPast ? months : -months));
			months = 0;
			remainingInterval += Math.abs(firstTime - changingDate.getTime()) / 1000;
		}
		return [years, months, remainingInterval];
	}

	if (!enableYears) { // No years but there must still be months
		const firstTime = changingDate.getTime();
		changingDate.setUTCMonth(changingDate.getUTCMonth() + (inPast ? -1 : 1));
		const monthDuration = Math.abs(firstTime - changingDate.getTime()) / 1000;
		if (remainingInterval >= monthDuration / 2) {
			months++;
		}
		months += years * 12;
		years = 0;
	} else if (!enableMonths) { // No months but there must still be years, so put the months back into interval
		const beforeMonths = changingDate.getTime();
		changingDate.setUTCMonth(changingDate.getUTCMonth() + (inPast ? months : -months));
		months = 0;
		remainingInterval += Math.abs(beforeMonths - changingDate.getTime()) / 1000;
		const firstTime = changingDate.getTime();
		changingDate.setUTCFullYear(changingDate.getUTCFullYear() + (inPast ? -1 : 1));
		const yearDuration = Math.abs(firstTime - changingDate.getTime()) / 1000;
		if (remainingInterval >= yearDuration / 2) {
			years++;
		}
	} else { // Both months and years, so round months
		const firstTime = changingDate.getTime();
		changingDate.setUTCMonth(changingDate.getUTCMonth() + (inPast ? -1 : 1));
		const monthDuration = Math.abs(firstTime - changingDate.getTime()) / 1000;
		if (remainingInterval >= monthDuration / 2) {
			months++;
			if (months >= 12) {
				months -= 12;
				years++;
			}
		}
	}
	return [years, months, 0];
};

type TimeUnit = "years" | "months" | "weeks" | "days" | "hours" | "minutes" | "seconds";
type StringifyThresholdsFull = { [K in TimeUnit]: [number, number] };
export type StringifyThresholds = Partial<{ [K in TimeUnit]: StringifyThresholdsFull[K] | number | boolean }>;
type StringifyThresholds2 = Record<TimeUnit, [number, number] | boolean>;

const thresholdDefaults: StringifyThresholdsFull = {
	years: [0, Infinity],
	months: [0, Infinity],
	weeks: [Infinity, 0],
	days: [0, Infinity],
	hours: [0, Infinity],
	minutes: [0, Infinity],
	seconds: [0, 10 * MINUTE_S]
};

type OtherStrings = {
	/** What is put between a number and its unit. Default: " " */
	spacer: string;
	/** What is put between any two units, except the penultimate and ultimate ones. Default: ", " */
	joiner: string;
	/** What is put between the penultimate and ultimate units. Default: " and " */
	finalJoiner: string;
};

type StringSettingsFull = Record<TimeUnit, [string, string]> & OtherStrings;
export type StringSettings = Partial<Record<TimeUnit, [string, string] | string> & OtherStrings>;

const stringDefaults: StringSettingsFull = {
	years: ["year", "years"],
	months: ["month", "months"],
	weeks: ["week", "weeks"],
	days: ["day", "days"],
	hours: ["hour", "hours"],
	minutes: ["minute", "minutes"],
	seconds: ["second", "seconds"],
	spacer: " ",
	joiner: ", ",
	finalJoiner: " and ",
};

type TimeUnitBooleansFull = Record<TimeUnit, boolean>;
/** The time units with booleans. Meaning depends on context. */
export type TimeUnitBooleans = Partial<TimeUnitBooleansFull>;

const timeUnitBooleanDefaults: TimeUnitBooleansFull = {
	years: false,
	months: false,
	weeks: false,
	days: false,
	hours: false,
	minutes: false,
	seconds: false,
};
/** All the stringifier options  */
export interface StringifyOptions {
	/** The date to use as a starting point, if months and years are desired */
	startDate?: Date;
	/**
	 * The thresholds for at which total intervals to display each unit.
	 * 
	 * Tuples indicate lower and upper thresholds. A single number will be used as an upper threshold.
	 * 
	 * `true` means `[0, Infinity]` (essentially enabling), and `false` means `[Infinity, 0]` (essentially disabling).
	 */
	thresholds?: StringifyThresholds;
	/**
	 * Whether to pad any given unit's number with 0s, like "5" -> "05".
	 * 
	 * Years will be padded to 4 digits, everything else to 2.
	 * 
	 * `true` enables all, `false` disables all. Default: `false`
	 */
	pad?: TimeUnitBooleans | boolean;
	/**
	 * Whether to still display units with a value of 0.
	 * 
	 * `true` enables all, `false` disables all. Default: `false`
	 */
	displayZero?: TimeUnitBooleans | boolean;
	/**
	 * Overrides for all the strings used in composing the final string.
	 * 
	 * Tuples on time units are used as singular and plural. A string will use the same for both.
	 */
	strings?: StringSettings;
}

/** StringifyOptions, but without startDate */
export type StringifierOptions = Omit<StringifyOptions, "startDate">;

/** All the stringifier options, except startDate, processed into a full set of settings */
interface StringifySettings {
	/**
	 * The thresholds for at which total intervals to display each unit.
	 * 
	 * Tuples indicate lower and upper thresholds.
	 */
	thresholds: StringifyThresholdsFull;
	/**
	 * Whether to pad any given unit's number with 0s, like "5" -> "05".
	 * 
	 * Years will be padded to 4 digits, everything else to 2.
	 */
	pad: TimeUnitBooleansFull;
	/**
	 * Whether to still display units with a value of 0.
	 */
	displayZero: TimeUnitBooleansFull;
	/**
	 * Overrides for all the strings used in composing the final string.
	 * 
	 * Tuples on time units are used as singular and plural.
	 */
	strings: StringSettingsFull;
}

const settingsDefaults: StringifySettings = {
	thresholds: thresholdDefaults,
	pad: timeUnitBooleanDefaults,
	displayZero: timeUnitBooleanDefaults,
	strings: stringDefaults,
};

type Entry<O, K extends keyof O> = [K, O[K]]
type Entries<O> = Entry<O, keyof O>[]

/** Processes a date or StringifyOptions into a StringifyOptions with all the defaults filled in */
const processStringifyOptions = (options: Date | StringifyOptions): { startDate?: Date; settings: StringifySettings } => {
	if (options instanceof Date) {
		const startDate = options;
		const settings = settingsDefaults;
		return { startDate, settings };
	}
	const { startDate, thresholds, pad, displayZero, strings } = options;
	const settings = {
		thresholds: processThresholdSettings(thresholds, startDate !== undefined),
		pad: processTimeUnitBooleans(pad),
		displayZero: processTimeUnitBooleans(displayZero),
		strings: processStringSettings(strings),
	};
	return { startDate, settings };
};

const timeElements: TimeUnit[] = ["years", "months", "weeks", "days", "hours", "minutes", "seconds"];

const processThresholdSettings = (settings: StringifyThresholds | undefined, date: boolean): StringifyThresholdsFull => {
	if (!settings)
		return thresholdDefaults;
	const {
		years = [0, Infinity],
		months = [0, Infinity],
		weeks = [Infinity, 0],
		days = [0, Infinity],
		hours = [0, Infinity],
		minutes = [0, Infinity],
		seconds = [0, 10 * MINUTE_S]
	} = settings;
	const finalSettings = { years, months, weeks, days, hours, minutes, seconds };
	for (const element of timeElements)
		finalSettings[element] = thresholdToTuple(finalSettings[element]);
	return finalSettings as StringifyThresholdsFull;
};

const thresholdToTuple = (input: boolean | number | [number, number]): [number, number] => {
	if (input === false)
		return [Infinity, 0];
	if (input === true)
		return [0, Infinity];
	if (typeof input === "number")
		return [0, input];
	return input;
};

const processTimeUnitBooleans = (booleans: TimeUnitBooleans | boolean | undefined): TimeUnitBooleansFull => {
	if (booleans === undefined || booleans === false)
		return timeUnitBooleanDefaults;
	if (booleans === true)
		return {
			years: true,
			months: true,
			weeks: true,
			days: true,
			hours: true,
			minutes: true,
			seconds: true,
		};
	const output: TimeUnitBooleans = {};
	for (const element of timeElements)
		output[element] = booleans[element] ?? timeUnitBooleanDefaults[element];
	return output as TimeUnitBooleansFull;
};

const processStringSettings = (settings: StringSettings | undefined): StringSettingsFull => {
	if (settings === undefined)
		return stringDefaults;
	const {
		years = ["year", "years"],
		months = ["month", "months"],
		weeks = ["week", "weeks"],
		days = ["day", "days"],
		hours = ["hour", "hours"],
		minutes = ["minute", "minutes"],
		seconds = ["second", "seconds"],
		spacer = " ",
		joiner = ", ",
		finalJoiner = " and "
	} = settings;
	const finalSettings = { years, months, weeks, days, hours, minutes, seconds, spacer, joiner, finalJoiner };
	for (const element of timeElements)
		finalSettings[element] = stringToTuple(finalSettings[element]);
	return finalSettings as StringSettingsFull;
};

const stringToTuple = (stringOrTuple: string | [string, string]): [string, string] => {
	return typeof stringOrTuple === "string" ? [stringOrTuple, stringOrTuple] : stringOrTuple;
};

/** Finds the smallest enabled constant unit to round interval to, and true if none were enabled */
const roundToSmallest = (interval: number, enabled: Record<TimeUnit, boolean>): [number, boolean?] => {
	if (enabled.seconds)
		return [interval];
	if (enabled.minutes)
		return [Math.round(interval / MINUTE_S) * MINUTE_S];
	if (enabled.hours)
		return [Math.round(interval / HOUR_S) * HOUR_S];
	if (enabled.days)
		return [Math.round(interval / DAY_S) * DAY_S];
	if (enabled.weeks)
		return [Math.round(interval / WEEK_S) * WEEK_S];
	return [interval, true];
};